ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision/BulkSync.hs

```diff
@@ -0,0 +1,613 @@
+                -- We chose the same peer as before but there were no more
+                -- blocks in flight for this peer: record a new start time.
+                peersOrderStart <- getMonotonicTime
+                writePeersOrder $ PeersOrder {peersOrderCurrent, peersOrderStart, peersOrderOthers}
+          _ -> pure ()
+
+      peerHasBlocksInFlight peer =
+        case find (\(_, (_, _, _, peer', _)) -> peer == peer') candidatesAndPeers of
+          Just (_, (_, inflight, _, _, _)) -> not $ Map.null $ peerFetchBlocksInFlight inflight
+          Nothing -> error "blocksInFlightForPeer"
+
+snoc :: [a] -> a -> [a]
+snoc [] a = [a]
+snoc (x : xs) a = x : snoc xs a
+
+-- | Given a list of candidate fragments and their associated peers, choose what
+-- to sync from who in the bulk sync mode.
+fetchDecisionsBulkSync ::
+  ( HasHeader header,
+    HeaderHash header ~ HeaderHash block,
+    Eq peer
+  ) =>
+  FetchDecisionPolicy header ->
+  -- | The current chain, anchored at the immutable tip.
+  AnchoredFragment header ->
+  (Point block -> Bool) ->
+  MaxSlotNo ->
+  PeersOrder peer ->
+  -- | Association list of the candidate fragments and their associated peers.
+  -- The candidate fragments are anchored in the current chain (not necessarily
+  -- at the tip; and not necessarily forking off immediately).
+  [(AnchoredFragment header, PeerInfo header peer extra)] ->
+  -- | Association list of the requests and their associated peers. There is at
+  -- most one accepted request; everything else is declined. Morally, this is a
+  -- map from peers to @'FetchDecision' ('FetchRequest' header)@ with at most
+  -- one @'FetchRequest' header@.
+  ( Maybe (FetchRequest header, PeerInfo header peer extra),
+    [(FetchDecline, PeerInfo header peer extra)]
+  )
+fetchDecisionsBulkSync
+  fetchDecisionPolicy
+  currentChain
+  fetchedBlocks
+  fetchedMaxSlotNo
+  peersOrder
+  candidatesAndPeers = combineWithDeclined $ do
+    -- Step 1: Select the candidate to sync from. This already eliminates peers
+    -- that have an implausible candidate. It returns the remaining candidates
+    -- (with their corresponding peer) as suffixes of the immutable tip.
+    ( theCandidate :: ChainSuffix header,
+      candidatesAndPeers' :: [(ChainSuffix header, PeerInfo header peer extra)]
+      ) <-
+      MaybeT $
+        selectTheCandidate
+          fetchDecisionPolicy
+          currentChain
+          candidatesAndPeers
+
+    -- Step 2: Filter out from the chosen candidate fragment the blocks that
+    -- have already been downloaded, or that have a request in flight (except
+    -- for the requests in flight that are ignored). NOTE: if not declined,
+    -- @theFragments@ is guaranteed to be non-empty.
+    let (theFragments :: FetchDecision (CandidateFragments header)) =
+          pure theCandidate
+            >>= filterNotAlreadyFetched fetchedBlocks fetchedMaxSlotNo
+            >>= filterNotAlreadyInFlightWithAnyPeerNonIgnored candidatesAndPeers
```

Comment:
Here's something else that puzzles me. I think the following claims are true:
1. All the in-flight requests of the current peer aren't ignored. The implementation does avoid rerequesting these in-flights.
2. The inflights of all other peers are ignored, as they have been rotated out of the best position in the `peersOrder`.

Therefore, first question: Why are we marking the blocks in inflight requests with a boolean in `PeerFetchBlockInFlight`? We know whether they are ignored or not by looking at which peer is at the front of the `peersOrder`.

Second question: Why bother dropping the inflight requests when preparing to select the peer? Surely we want to select a peer that can offer all the blocks that haven't been downloaded yet.

Reply:
I think you're right and we could (should) definitely get rid of the boolean tracking ignored blocks. This would involve changing the way we track the current peer. Currently, it is by definition the only one that has non-ignored in-flight requests. However, we could just add it to a TVar somewhere, erase it on starvations, and say that the current peer is that one _if_ it has requests in flights, and otherwise we may elect a new one. I think this would be much cleaner, and it would drastically simplify the specification by removing this notion of ignored vs. non-ignored in-flight blocks.

Reply:
@nfrisby> Are we missing something here?

Reply:
We got rid of the notion of ignored in-flight requests.